renku/service/controllers/migrations_check.py

```diff
@@ -0,0 +1,61 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2020 - Swiss Data Science Center (SDSC)
+# A partnership between École Polytechnique Fédérale de Lausanne (EPFL) and
+# Eidgenössische Technische Hochschule Zürich (ETHZ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Renku service migrations check controller."""
+from renku.core.commands.migrate import migrations_check, migrations_versions
+
+from renku.service.controllers.mixins import ReadOperationMixin
+
+from renku.service.serializers.cache import ProjectMigrationCheckRequest, ProjectMigrationCheckResponseRPC
+from renku.service.views import result_response
+
+
+class MigrationsCheckCtrl(ReadOperationMixin):
+    """Controller for migrations check endpoint."""
+
+    REQUEST_SERIALIZER = ProjectMigrationCheckRequest()
+    RESPONSE_SERIALIZER = ProjectMigrationCheckResponseRPC()
+
+    def __init__(self, cache, user_data, request_data):
+        """Construct migration check controller."""
+        self.ctx = MigrationsCheckCtrl.REQUEST_SERIALIZER.load(request_data)
+        super(MigrationsCheckCtrl, self).__init__(cache, user_data, request_data)
+
+    @property
+    def context(self):
+        """Controller operation context."""
+        return self.ctx
+
+    def renku_op(self):
+        """Renku operation for the controller."""
+        latest_version, project_version = migrations_versions()
+        migration_required, project_supported = migrations_check()
+
+        return {
+            "migration_required": migration_required,
+            "project_supported": project_supported,
+            "project_version": project_version,
+            "latest_version": latest_version,
+        }
+
+    def to_response(self):
+        """Execute controller flow and serialize to service response."""
+        if "project_id" in self.ctx:
+            return result_response(MigrationsCheckCtrl.RESPONSE_SERIALIZER, self.local())
+
+        elif "git_url" in self.ctx:
+            return result_response(MigrationsCheckCtrl.RESPONSE_SERIALIZER, self.remote())
```

Comment:
should this also return `[]` like the other Ctrls?

Reply:
What do you mean? `result_response` will always return JSON-RPC structure response, the `result` part of this structure will always be determined by the controller response serializers and the data for the response will come from local or remote execution flow.

Reply:
the other two controls have a structure like 
```
        self.ctx["..."] = []

        if "project_id" in self.ctx:
            self.ctx["..."] = self.local()

        elif "git_url" in self.ctx:
            self.ctx["..."] = self.remote()

        return result_response(<serializer>, self.ctx)
```

And this control has essentially the same functionality inside the if/elif branches but does not have the `self.ctx["..."] = []` default value, so I wondered why the default value in the other two ctrls was needed but not here, since the structure is virtually identical.